File: src/cied_robotic_arm_moveit_servo/scripts/joy_test.py
import rclpy
from rclpy.node import Node
from sensor_msgs.msg import Joy
from geometry_msgs.msg import TwistStamped
from std_msgs.msg import String

class GamepadController(Node):
    def __init__(self):
        super().__init__('gamepad_controller')
        self.joy_sub = self.create_subscription(Joy, '/joy', self.joy_callback, 10)
        self.servo_pub = self.create_publisher(TwistStamped, '/servo_commands', 10)
        self.mode = 'joint_control'  # Default mode

    def joy_callback(self, msg):
        # Example mapping logic:
        if msg.buttons[0]:  # A button
            self.mode = 'joint_control'
            self.get_logger().info('Switched to Joint Control Mode')
        elif msg.buttons[1]:  # B button
            self.mode = 'cartesian_control'
            self.get_logger().info('Switched to Cartesian Control Mode')

        if self.mode == 'cartesian_control':
            twist = TwistStamped()
            twist.twist.linear.x = msg.axes[0]  # LeftX
            twist.twist.linear.y = msg.axes[1]  # LeftY
            twist.twist.angular.z = msg.axes[2]  # RightX
            self.servo_pub.publish(twist)

def main(args=None):
    rclpy.init(args=args)
    node = GamepadController()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/cied_robotic_arm_moveit_servo/src/joy_test.cpp
#include "rclcpp/rclcpp.hpp"
#include "sensor_msgs/msg/joy.hpp"
#include "control_msgs/msg/joint_jog.hpp"
#include "geometry_msgs/msg/twist_stamped.hpp"
#include "std_msgs/msg/string.hpp"

class GamepadController : public rclcpp::Node {
public:
    GamepadController() : Node("gamepad_controller"), mode_("joint_control") {
        joy_sub_ = this->create_subscription<sensor_msgs::msg::Joy>(
            "/joy", 10, std::bind(&GamepadController::joyCallback, this, std::placeholders::_1));

        twist_pub_ = this->create_publisher<geometry_msgs::msg::TwistStamped>(
            "/servo_server/delta_twist_cmds", 10);

        
        joint_cmds_pub_ = this->create_publisher<control_msgs::msg::JointJog>(
            "/servo_server/delta_joint_cmds", 10);

        RCLCPP_INFO(this->get_logger(), "Gamepad controller node initialized.");
    }

private:
    void joyCallback(const sensor_msgs::msg::Joy::SharedPtr msg) {
        // Mode switching logic
        if (msg->buttons[0] == 1) { // A button
            mode_ = "joint_control";
            RCLCPP_INFO(this->get_logger(), "Switched to Joint Control Mode");
        } else if (msg->buttons[1] == 1) { // B button
            mode_ = "cartesian_control";
            RCLCPP_INFO(this->get_logger(), "Switched to Cartesian Control Mode");
        }

        geometry_msgs::msg::TwistStamped twist_msg;
        twist_msg.header.stamp = this->get_clock()->now();

        // Cartesian Control Mode
        if (mode_ == "cartesian_control") {
            twist_msg.twist.linear.x = msg->axes[0];  // Left stick X for x movement
            twist_msg.twist.linear.y = msg->axes[1];  // Left stick Y for y movement
            twist_msg.twist.linear.z = (msg->buttons[3] == 1) ? 0.5 : (msg->buttons[4] == 1 ? -0.5 : 0.0);  // Z-axis control with X/Y buttons
            twist_msg.twist.angular.z = msg->axes[2];  // Right stick X for rotation

            twist_pub_->publish(twist_msg);
        }

        // Future expansion for joint control mode
        if (mode_ == "joint_control") {
            control_msgs::msg::JointJog joint_cmds;
            joint_cmds.header.stamp = this->get_clock()->now();

            // // Example mapping of joystick axes to joint commands
            // joint_cmds.velocities[0] = msg->axes[4];  // Left trigger for joint 1
            // joint_cmds.velocities[1] = msg->axes[5];  // Right trigger for joint 2
            // joint_cmds.velocities[2] = msg->axes[2];  // Left stick X for joint 3
            // joint_cmds.velocities[3] = msg->axes[3];  // Left stick Y for joint 4
            // joint_cmds.velocities[4] = msg->axes[0];  // Right stick X for joint 5
        }
    }

    std::string mode_;
    rclcpp::Subscription<sensor_msgs::msg::Joy>::SharedPtr joy_sub_;
    rclcpp::Publisher<geometry_msgs::msg::TwistStamped>::SharedPtr twist_pub_;
    rclcpp::Publisher<control_msgs::msg::JointJog>::SharedPtr joint_cmds_pub_;
};

int main(int argc, char **argv) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<GamepadController>());
    rclcpp::shutdown();
    return 0;
}
